bin/oneoff/backfill_data/backfill_script_ids_channel_tokens.rb

```diff
@@ -0,0 +1,211 @@
+          update_channel_token(channel_token, script_id, csv)
+          next
+        end
+
+        # it's possible the channel token was generated for a logged-out user, in which case no user_id will exist
+        if user_id.present?
+          # get user_scripts for the associated_script_ids, if there is just one user_script, backfill with the
+          # associated script_id
+          script_ids_from_user_scripts = associated_script_ids_by_user_scripts(user_id, associated_script_ids)
+          if script_ids_from_user_scripts.count == 1
+            update_channel_token(channel_token, script_ids_from_user_scripts[0], csv)
+            next
+          end
+
+          # get user_levels for the associated_script_ids, if there is just one user_level, backfill with the
+          # associated script_id
+          script_ids_from_user_levels = associated_script_ids_from_user_levels(user_id, level)
+          if script_ids_from_user_levels.count == 1
+            update_channel_token(channel_token, script_ids_from_user_levels[0], csv)
+            next
+          end
+
+          # if the user has no associated user_levels or user_scripts, it's possible that the user visited a channel backed
+          # level which generated a channel_token and then the user left the page without making any progress. In this case it doesn't
+          # matter which script_id we pick, so backfill with the first associated_script_ids
+          if associated_script_ids.count > 0 && script_ids_from_user_scripts.blank? && script_ids_from_user_levels.blank?
+            update_channel_token(channel_token, associated_script_ids[0], csv)
+            next
+          end
+        end
+      rescue StandardError
+        print "[E]"
+      end
+
+      log_backfill_failed(channel_token.id, csv)
+    end
+  end
+  puts
+  puts "Script ended at #{Time.now}"
+  puts
+  puts "unable to backfill script id for #{$unable_to_backfill} channel tokens"
+  puts "Failures exported to: #{$csv_filename}"
+  puts
+end
+
+def get_associated_script_ids(level)
+  if $level_to_script_ids[level.id].present?
+    return $level_to_script_ids[level.id]
+  end
+
+  script_ids = level.script_levels.map(&:script_id)
+
+  level.parent_levels.map do |parent_level|
+    parent_level_script_ids = parent_level.script_levels.map(&:script_id)
+    script_ids.concat(parent_level_script_ids)
+  end
+
+  $level_to_script_ids[level.id] = script_ids.uniq
+  return $level_to_script_ids[level.id]
+end
+
+def update_channel_token(channel_token, script_id, csv)
+  unless $is_dry_run
+    did_save = channel_token.update_attributes(script_id: script_id)
+
+    unless did_save
+      log_backfill_failed(channel_token.id, csv)
+      return
+    end
+  end
+  print "#{channel_token.id},"
+end
+
+def log_backfill_failed(channel_token_id, csv)
+  print "[F] #{channel_token_id},"
+  $unable_to_backfill += 1
+  csv << [channel_token_id]
+end
+
+def associated_script_ids_by_user_scripts(user_id, script_ids)
```

Comment:
(minor) Rename this to `associated_script_ids_from_user_scripts` to be consistent with the method for user_levels?